skimage/util/lazy.py

```diff
@@ -0,0 +1,52 @@
+import sys
+import importlib
+import importlib.util
+
+
+def install_lazy(module_name, submodules=None, submod_attrs=None):
+    """Install lazily loaded submodules, and functions or other attributes.
+
+    Parameters
+    ----------
+    module_name : str
+        Typically use __name__.
+    submodules : set
+        List of submodules to install.
+    submod_attrs : dict
+        Dictionary of submodule -> list of attributes / functions.
+        These attributes are imported as they are used.
+
+    Returns
+    -------
+    __getattr__, __dir__, __all__
+
+    """
+    if submod_attrs is None:
+        submod_attrs = {}
+
+    if submodules is None:
+        submodules = set()
+    else:
+        submodules = set(submodules)
+
+    attr_to_modules = {
+        attr: mod for mod, attrs in submod_attrs.items() for attr in attrs
+    }
+
+    __all__ = submodules | attr_to_modules.keys()
+
+    def __getattr__(name):
+        if name in submodules:
+            return importlib.import_module(f'{module_name}.{name}')
```

Comment:
In a previous version of this there was a "require" logic. It did stuff like `importlib.util.find_spec` and making `importlib.util.LazyLoader(spec.loader)`. I'm not sure if that is necessary or not. If its not necessary, great, I like getting rid of superfluous code, but I'm curious why it was there to begin with and what made you remove it?

Reply:
I think that was primarily because of the evolution of the idea—I started with lazy module imports, and then added the other layers on top.

With `__dir__` in place, you can inspect the module, and only on `__getattr__` should the module be loaded.  The loaded module has the same setup, so it won't be expensive to import either.

You have a keen eye, I must say!